### PR TITLE
Don't try to bind on address from inactive interface

### LINKED ISSUE #934: Consul is trying to bind to an interface that isn't up
I was having a problem where the first address being found was for an interface that was in the "down" state.

In consul/util.go:GetPrivateIP() you should probably be using net.Interfaces() and then select an interface that has FlagUp set and also does NOT have FlagLoopback set to make it more robust in finding a usable private IP.

### PR DESCRIPTION
**Pull request by tiwilliam**
Build our own network address list, we now make sure the interface is up and supports both broadcast and multicast. Resolves #934.

### CODE REVIEW

**Comment by tiwilliam** [comment]
Will probably resolve #928 as well.

**Comment by tiwilliam** [comment]
We seem to rely on resolving `127.0.0.1` in tests, maybe we want to bind to loopback if nothing else can be found.

File: consul/util.go [hunk 1]
```diff
@@ -153,10 +153,37 @@ func isPrivateIP(ip_str string) bool {
...
+// both broadcast and multicast
+func activeInterfaceAddresses() ([]net.Addr, error) {
+	var out []net.Addr
+
+	interfaces, err := net.Interfaces()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get interfaces: %v", err)
+	}
+
+	reqFlags := net.FlagUp | net.FlagBroadcast | net.FlagMulticast
```
Comment (armon) [inline consul/util.go]: We don't actually require Broadcast or Multicast support. Any reason to check for it? Also how does this method affect Windows?

Comment (tiwilliam) [inline consul/util.go]: Reasoning for filtering on broadcast and multicast was to filter out loopback and other dummy interfaces, could change that to be a explicit filter on `FlagLoopback` instead.

Windows support all these flags, see https://golang.org/src/net/interface_windows.go.

**Comment by tiwilliam** [comment]
Alright, we will now only require interfaces to be up and use loopback if nothing else can be found.

**Comment by armon** [comment]
@tiwilliam My only concern now is Windows. I have a suspicion it will not play nice...

**Comment by tiwilliam** [comment]
Alright, let me find some testing units and play around.

**Comment by tiwilliam** [comment]
Someone with easy access to Windows, feel free to test this out. I've not gathered energy enough for installing Windows in my VM.

**Comment by highlyunavailable** [comment]
Works fine for me on Windows:

```
==> WARNING: Bootstrap mode enabled! Do not enable unless necessary
==> WARNING: Windows is not recommended as a Consul server. Do not use in production.
==> WARNING: It is highly recommended to set GOMAXPROCS higher than 1
==> Starting Consul agent...
==> Starting Consul agent RPC...
==> Consul agent running!
         Node name: 'MY-PC'
        Datacenter: 'dc1'
            Server: true (bootstrap: true)
       Client Addr: 127.0.0.1 (HTTP: 8500, HTTPS: -1, DNS: 8600, RPC: 8400)
      Cluster Addr: 10.0.240.1 (LAN: 8301, WAN: 8302)
    Gossip encrypt: false, RPC-TLS: false, TLS-Incoming: false
             Atlas: <disabled>

==> Log data will now stream in as it occurs:

    2015/06/29 14:09:00 [INFO] serf: EventMemberJoin: MY-PC 10.0.240.1
```

This was with 2 virtualbox adapters (enabled) and my LAN adapter disabled.

**Comment by highlyunavailable** [comment]
Correction: I get this error on Windows if I try to bind with all 3 adapters disabled:

```
==> Starting Consul agent...
==> Error starting agent: Failed to get advertise address: Failed to get interface addresses: Failed to get interfaces: GetAdaptersInfo: The pipe is being closed.
```

**Comment by tiwilliam** [comment]
@highlyunavailable Thank you! I wonder if it works with the Windows Loopback Adapter installed.

**Comment by highlyunavailable** [comment]
No, it fails with `==> Error starting agent: Failed to get advertise address: No private IP address found`.

However, this is because [privateBlocks](https://github.com/hashicorp/consul/blob/master/consul/util.go) does not contain 169.254.0.0/16, and that's what the loopback adapter binds to since it can't find another IP. Interestingly, that file also does not contain 127.0.0.0/8, so you might run into a problem there.

**Comment by tiwilliam** [comment]
@highlyunavailable Good catch, I've added `127.0.0.0/8` and we now treat it as a private block, which I think in this case is perfectly fine.

I guess you can change address of your Windows Loopback Adapter manually to 127.0.0.1 if you really want that?

**Comment by highlyunavailable** [comment]
You can't, because 127.0.0.1 would get added as part of the adapter. I strongly suggest adding 169.254.0.0/16 anyway, since it is defined as a private range in [RFC 3927](https://tools.ietf.org/html/rfc3927).

**Comment by tiwilliam** [comment]
Alright, makes sense. My only concern would be that we now might end up using a non-working DHCP interface if that's first in the interface list.

**Comment by highlyunavailable** [comment]
True. I'd say if you're going to support 169.254 then it should be the absolute 100% last thing to bind to - all other non-loopback adapters should be tried first.

**Comment by tiwilliam** [comment]
I'm happy with current state of this now when it's tested on Windows. A final sign-off before merge would be appreciated.

**Comment by highlyunavailable** [comment]
It doesn't look like the 169.254 case is a problem since Consul will [bail out](https://github.com/hashicorp/consul/blob/iface-down-fix/consul/util.go#L255) if it is presented with multiple private IPs.

This could only be a problem if the machine a. has 1 APIPA-addressed DHCP adapter b. public IP adapter(s) - this would mean that consul would try to auto-bind to the APIPA adapter, but that's already a thing - Consul won't bind to a public IP. I'm find with how it works in Windows right now as well.

if @armon approves it would be cool to get this merged - there are multiple open bugs that this would fix it looks like: #934, #928.

**Comment by slackpad** [comment]
Sorry for the delay on this one - looks good!

**Pull request merged**